src/servers/tracer.cc

```diff
@@ -242,4 +246,215 @@ TraceManager::TraceActivity(
       << "\",\"ns\":" << timestamp_ns << "}]}";
 }
 
+void
+TraceManager::TraceTensorActivity(
+    TRITONSERVER_InferenceTrace* trace,
+    TRITONSERVER_InferenceTraceActivity activity, const char* name,
+    TRITONSERVER_DataType datatype, const void* base, size_t byte_size,
+    const int64_t* shape, uint64_t dim_count,
+    TRITONSERVER_MemoryType memory_type, int64_t memory_type_id, void* userp)
+{
+  if ((activity != TRITONSERVER_TRACE_TENSOR_INPUT) &&
+      (activity != TRITONSERVER_TRACE_TENSOR_OUTPUT)) {
+    return;
+  }
+
+  void* buffer_base = nullptr;
+  if (memory_type == TRITONSERVER_MEMORY_GPU) {
```

Comment:
What is the design plan for how tensor data is handled? Above we discussed using  CopyBuffer but why are you copying within core Triton at all. If you are synchronously calling this function with 'base', 'memory_type' and 'memory_type_id' then you should just do all the copying here to get the tensor data into CPU memory so that you can output it.